exonum/src/storage/proof_list_index/mod.rs

```diff
@@ -59,17 +65,28 @@ pub struct ProofListIndexIter<'a, V> {
     base_iter: BaseIndexIter<'a, ProofListKey, V>,
 }
 
+/// TBD
+pub fn hash_leaf<V: StorageValue>(value: V) -> Hash {
+    let value_bytes = value.into_bytes();
+    hash_with_prefix(LEAF_TAG, &value_bytes)
+}
+
 fn hash_one(h: &Hash) -> Hash {
-    hash(h.as_ref())
+    hash_with_prefix(NODE_TAG, h.as_ref())
 }
 
 fn hash_pair(h1: &Hash, h2: &Hash) -> Hash {
     HashStream::new()
+        .update(&[NODE_TAG])
         .update(h1.as_ref())
         .update(h2.as_ref())
         .hash()
 }
 
+pub fn hash_with_prefix(prefix: u8, value: &[u8]) -> Hash {
+    HashStream::new().update(&[prefix]).update(&value).hash()
```

Comment:
If I didn't know that `HashStream` is slow (see discussion in #1046 for details), I'd suggest to store `HashStream::new().update(&[prefix])` for allowed `prefix`es e.g. as `lazy_static` in order to speed up calculations. But I'm not sure they would be sped up at all 🙃 Also, as far as I can see, `HashStream` doesn't implement `Clone` and is passed by value for each `update`, which makes use of `lazy_static` non-trivial. (Don't know whether this is a problem in Rust binding that can be fixed.)

Reply:
If we care for it that much then I'd rather look pinpoint that performance issue in `HashStream` implementation and fix it. Otherwise leave this code as is.

Even if we do care to avoid using `HashStream`, I think that doing

```rust
let mut value = value.to_vec();
value.insert(0, prefix);
```

will be much more concise and readable than any trickery with `lazy_static` and partially-computed hashes.

It's _a single-byte prefix_ and we're computing a hash sum for multi-byte value. Allocating a new hash state and populating it with a single byte should not be significantly more expensive than cloning an existing hash state.

Reply:
Sorry for my brainfart; caching hash state most likely won't do anything good, as SHA-256 is processed in chunks of 64 bytes. Still, I'd run benchmarks and would not use `HashStream` if it turns out to be slow. I have a suspicion that `Vec::insert` may be slow, too; creating a vector with sufficient length from the start and then filling it could be faster (in some cases, we can use an array, as we know length of `value`).

Reply:
Bench is [here](https://github.com/pavel-mukhanov/sandbox/blob/master/src/proof.rs)